### PR TITLE
Gemini channels list 🏷 

### PR DESCRIPTION
**Pull request by dlipeles**
## Gemini channels list

Screen Shot 2019-06-11 at 6 48 16 PM

#### Important 🔑 

- We need to notify publishers via an e-mail about these changes before we take this to production
- We should consider upgrading the redis cache since we'll be storing v1 and v2 lists

#### Features

* Supports Gemini changes requiring publisher KYC & wallet address
* Creates new `/api/v2/public/channels` endpoint for Gemini channels list
* Heavy refactor of `channels_json_builder.rb` with more modularity
* Removes extraneous logic that included excluded verified channels. 
* Backwards compat. with `api/v1/public/channels` endpoint

#### Notes

This feature contains the changes for Gemini, but I also took the opportunity to refactor this file as it is important to Publisher/Browser operations. I put much of the logic into modular functions, which should help maintenance / readability a lot. 

I also noticed there was some existing logic that was 
1. Including excluded but _verified_ channels along with the rest of the channels.
2. Appending excluded but _not verified_ channels at the end of the list. 

I couldn't see a reason why we should include excluded channels with the others even if they are verified, so I removed that logic. Now all the excluded channels are appended to the end of the list. 

Lastly I moved the site banners logic to the `publishers_helper.rb` file, to simplify the json builder file. This will also make a future refactor (where we omit default site banners from the list) easier. 

#### How To Test

1. Create a publisher and verify a few channels 
2. Check the channels list at `https://localhost:3000/api/v1/public/channels`
3. Ensure that the channels do not appear (because no KYC'd Uphold yet) 
4. Add an Uphold Account (preferably without KYC) 
5. Ensure that the channels do not appear (because no KYC'd Uphold yet) 
6. KYC the uphold connection
7. Ensure that the channels appear now, along with the wallet id

#### Pull Request Checklist

- [x] Adequate test coverage exists to prevent regressions -- [Guide to Testing](https://guides.rubyonrails.org/testing.html)
- [x] XSS is mitigated -- [Guide to XSS Prevention](https://guides.rubyonrails.org/security.html#cross-site-scripting-xss)
- [x] No raw SQL -- [Guide to SQL Injection Prevention](https://guides.rubyonrails.org/security.html#sql-injection)
- [x] UI/UX is responsive -- [Guide to Responsive UI/UX](https://developers.google.com/web/fundamentals/design-and-ux/responsive/)
- [x] Integrated Matomo for new UI elements -- [Guide to Matomo](https://developer.matomo.org/guides/integrate-introduction)
- [x] Passes checklist for Progressive Web App -- [Guide to PWAs](https://developers.google.com/web/progressive-web-apps/checklist)

### CODE REVIEW

File: app/helpers/publishers_helper.rb [hunk 1]
```diff
@@ -479,4 +479,15 @@ def youtube_login_permitted?(channel)
...
   end
+
+  def site_banner_details(channel)
+    publisher = channel.publisher
+    if publisher.default_site_banner_mode && publisher.default_site_banner_id
+      publisher.default_site_banner.read_only_react_property
+    elsif channel.site_banner
+      channel.site_banner.read_only_react_property
+    else
+      {}
```
[the remark this answers is not in the document]
Comment (NejcZdovc) [reply app/helpers/publishers_helper.rb]: yup we check if get this object, so we could remove it and reduce size. If we don't find or is empty  it we will just use default

[the remark this answers is not in the document]
Comment (dlipeles) [reply app/helpers/publishers_helper.rb]: Yup see the last line I wrote under the `notes` section in the PR desc.

File: app/helpers/publishers_helper.rb [hunk 2]
```diff
@@ -479,4 +479,15 @@ def youtube_login_permitted?(channel)
   def publisher_created_through_youtube_auth?(publisher)
     publisher && publisher.channels.visible.count == 1 && youtube_login_permitted?(publisher.channels.visible.first)
   end
+
+  def site_banner_details(channel)
```
[the remark this answers is not in the document]
Comment (dlipeles) [reply app/helpers/publishers_helper.rb]: Sure, good feedback

File: app/models/json_builders/channels_json_builder.rb [hunk 3]
```diff
@@ -16,44 +17,69 @@
...
+  ############################################################################################
+  # Step three, if channel not on exclusion list, and publisher has KYC'd, add the channel.
+  ############################################################################################
+
+  @channels.push([
+    verified_channel.details.channel_identifier, 
+    true, 
+    false, 
+    site_banner_details(verified_channel), 
+    verified_channel.publisher.uphold_connection.address
```
[the remark this answers is not in the document]
Comment (dlipeles) [reply app/models/json_builders/channels_json_builder.rb]: I put the nil-safe on 

```ruby
    return if verified_channel.publisher&.uphold_connection&.is_member != true
```

which should return if anything is nil

File: app/models/json_builders/channels_json_builder.rb [hunk 4]
```diff
@@ -16,44 +17,69 @@
...
-    [
-      Channel.verified.site_channels.includes(:site_banner).includes(publisher: :site_banners),
-      Channel.verified.youtube_channels.includes(:site_banner).includes(publisher: :site_banners),
-      Channel.verified.twitch_channels.includes(:site_banner).includes(publisher: :site_banners),
-      Channel.verified.twitter_channels.includes(:site_banner).includes(publisher: :site_banners),
-      Channel.verified.vimeo_channels.includes(:site_banner).includes(publisher: :site_banners),
-      Channel.verified.reddit_channels.includes(:site_banner).includes(publisher: :site_banners),
-      Channel.verified.github_channels.includes(:site_banner).includes(publisher: :site_banners),
-    ].each do |verified_channels|
+    joined_verified_channels = join_verified_channels
```
[the remark this answers is not in the document]
Comment (dlipeles) [reply app/models/json_builders/channels_json_builder.rb]: Did it for readability, compare 

```ruby
  def build
    joined_verified_channels = join_verified_channels
    joined_verified_channels.each do |verified_channels|
      verified_channels.find_each do |verified_channel|
        include_verified_channel(verified_channel) if @version == "v1"
        include_verified_channel_v2(verified_channel) if @version == "v2"
      end
    end
    append_excluded
    @channels.to_json
  end
```

to 

```ruby 
  def build
    channels = []
    [
      Channel.verified.site_channels.includes(:site_banner).includes(publisher: :site_banners),
      Channel.verified.youtube_channels.includes(:site_banner).includes(publisher: :site_banners),
      Channel.verified.twitch_channels.includes(:site_banner).includes(publisher: :site_banners),
      Channel.verified.twitter_channels.includes(:site_banner).includes(publisher: :site_banners),
      Channel.verified.vimeo_channels.includes(:site_banner).includes(publisher: :site_banners),
      Channel.verified.reddit_channels.includes(:site_banner).includes(publisher: :site_banners),
      Channel.verified.github_channels.includes(:site_banner).includes(publisher: :site_banners),
    ].each do |verified_channels|
      verified_channels.find_each do |verified_channel|
        if @excluded_channel_ids.include?(verified_channel.details.channel_identifier)
          excluded = true
          @excluded_verified_channel_ids.push(verified_channel.details.channel_identifier)
        else
          excluded = false
        end
        if verified_channel.publisher.default_site_banner_mode && verified_channel.publisher.default_site_banner_id
          channels.push([verified_channel.details.channel_identifier, true, excluded, verified_channel.publisher.default_site_banner.read_only_react_property])
        elsif verified_channel.site_banner
          channels.push([verified_channel.details.channel_identifier, true, excluded, verified_channel.site_banner.read_only_react_property])
        else
          channels.push([verified_channel.details.channel_identifier, true, excluded, {}])
        end
      end
    end
```

File: app/models/json_builders/channels_json_builder.rb [hunk 5]
```diff
@@ -20,10 +20,22 @@ def initialize
     require "publishers/excluded_channels"
     @excluded_channel_ids = Publishers::ExcludedChannels.brave_publisher_id_list
     @excluded_verified_channel_ids = []
+    @channels = []
```
Comment (yachtcaptain23) [inline app/models/json_builders/channels_json_builder.rb]: Would you revert all the changes in this file? Let me know if you want help

File: test/models/json_builders/channels_json_builder_test.rb [hunk 6]
```diff
@@ -53,21 +52,6 @@ def get_channel_from_json(channels, channel_id)
     refute channel
   end
 
-  test "verified channel that is excluded is returned and marked correctly" do
```
Comment (yachtcaptain23) [inline test/models/json_builders/channels_json_builder_test.rb]: Please also restore this test

**Pull request merged**